[DI] Rework config hierarchy: defaults > instanceof > service config

weaverryan opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no (because instanceof is new)
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | instanceof symfony/symfony-docs#7538

This reworks how `defaults`, `instanceof` and service config override each other. For example:

```yml
services:
    _defaults:
        autowire: true

    _instanceof:
        Symfony\Component\Security\Core\Authorization\Voter\VoterInterface:
            public: false
            autowire: false
            tags: [security.voter]

    AppBundle\Security\PostVoter:
        public: true
        tags:
            - { name: security.voter, priority: 100 }
```

|                       | Before | After | Changed?
| ------------- | ------------- | ---- | --- |
| **public**   | false <br/> (taken from `_instanceof`  | true<br/> (service overrides `_instanceof`) | **changed**
| **autowired**   | false  | false | no change |
| **tags**   | 2 `security.voter` tags  | 2 `security.voter` tags  | no change |
| **Logic**   | `_defaults` > service config > `_instanceof`   | `_defaults` > `_instanceof` > service config |

tl;dr; `_instanceof` now acts like `_defaults`, but for specific instances. As a user, this is how I expected it to work originally.

**NOTE** Much of the config merging behavior in this PR - i.e. `tags`, `calls` - could be accomplished without such a big diff. But any scalar config - i.e. `public`, `autowire`, `configurator`, etc - *requires* all of the changes in this PR.

Important Notes:

* `_instanceof` has its own override/merge logic. Mostly, it's easy - scalar values override. Some are more opinionated:
   * `tags` from `_instanceof` and service are merged - all tags are kept. It's up to the compiler pass to decide of 2 tags should be merged into 1 or if both are needed
   * `calls` from `_instanceof` and service are merged, unless the method name of the call match, then only the one from the service wins (i.e. ability to override a call). `_instanceof` calls are called *first*
* I removed support for `arguments`, `factory`, `deprecated` and `abstract` from `_instanceof`, and I think we could remove more. These feel like an abuse of `_instanceof` when a parent class is more proper (and the behavior of how parent-child config is merged is more predictable in general)
* This also works with parent-child definitions. The parent-child relationship is resolved first and then `instanceof` is applied. This means the real hierarchy is:

> `_defaults` > `_instanceof` > `parent definition` > `child definition`

And yep, there's a test for this in `IntegrationTest` :)

* [Bug Fix] instanceofConditionals were being lost when a service had a parent. This was fixed in sha: 62029090d35bb69d8ad4b5f6714592ac18ba9456. We simply take the conditionals from the child... which in practice will likely match the parent conditionals, unless they were created in different YAML files (in which case, using the `_instanceof` from the file where the child was configured is what we want)

Cheers!

weaverryan committed:
Reworking the instanceof feature to act more like a "default" than like a ChildDefinition

tl;dr Before, instanceof values would replace those explicitly set on services. Now, it
acts more like a default: only setting a value on a service if that value wasn't explicitly
set on that service (i.e. don't override). For things like tags and calls, they're merged
intelligently and on a case-by-case basis (merging tags works different than merging calls)

weaverryan commented on src/Symfony/Component/DependencyInjection/Compiler/PassConfig.php:
```diff
@@ -42,13 +42,13 @@ public function __construct()
         $this->beforeOptimizationPasses = array(
             100 => array(
                 $resolveClassPass = new ResolveClassPass(),
-                new ResolveDefinitionInheritancePass(),
```
I moved this down to be *after* `ResolveDefinitionTemplatesPass` on purpose: this allows the parent-child definitions to be resolved first. Then, the `ResolveDefinitionInheritancePass` doesn't need to worry about this.

nicolas-grekas replied:
Thinking about that, this should be kept here: all following compiler passes need to always win. Moving the pass after all DI extension passes is what makes changes tracking a nightmare.
unless I missed something, the full target order proposed in this PR should be:
defaults > instanceof > parents > service > passes

nicolas-grekas replied:
(vs defaults > parents > service > instanceof > passes as implemented today)

weaverryan commented on src/Symfony/Component/DependencyInjection/Compiler/ResolveDefinitionTemplatesPass.php:
```diff
@@ -102,26 +102,6 @@ private function doResolveDefinition(ChildDefinition $definition)
-     * @internal
-     */
-    public static function mergeDefinition(Definition $def, ChildDefinition $definition)
-    {
```
This class was simply changed to look like it did before #21530 (adjusted for changes since then)

weaverryan commented on src/Symfony/Component/DependencyInjection/Compiler/ResolveDefinitionInheritancePass.php:
```diff
@@ -27,7 +27,7 @@ protected function processValue($value, $isRoot = false)
         }
 
-        $class = $value instanceof ChildDefinition ? $this->resolveDefinition($value) : $value->getClass();
+        $class = $value->getClass();
```
This entire class/file could/should be renamed (`ResolveInstanceofPass`) - I didn't do it to keep the diff more manageable

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Definition.php:
```diff
@@ -803,4 +818,14 @@ public function hasAutowiringType($type)
+     *
+     * @return array An array of configured parts for this Definition
+     */
+    public function getConfiguredParts()
```
should this replace getChanges on ChildDefinition? (or have getChanges migrate here instead?)

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/ResolveDefinitionInheritancePass.php:
```diff
@@ -39,57 +39,50 @@ protected function processValue($value, $isRoot = false)
+        // merge tags
+        $tags = $def->getTags();
+        foreach ($instanceofDefinition->getTags() as $k => $v) {
+            // don't add a tag if one by that name was already added
```
that's cheating :)
this is unrelated to the other changes, and could be done on the existing code base already

nicolas-grekas commented:
> don't add a tag if one by that name was already added

this is already doable with a minor change in the existing code base, isn't it?
if yes, then what's the purpose of this PR? I mean: either should be updated, of the patch simplified drastically.
Or do I miss something?

stof commented:
@weaverryan We still must have instanceof winning over defaults. Otherwise, it removes the possibility to say "I want all services in this file to be private, except for commands because they are required to be public".

Service-level config should win over instanceof, but not service defaults.

stof commented:
the order needs to be ``_defaults > _instanceof > service``, whcih each step winning over the previous one.

weaverryan committed:
Moving changes tracking from ChildDefinition to Definition and fixing bug

The bug is where _defaults was overriding _instanceof config. To handle that,
we ignore change tracking when setting from _defaults

weaverryan committed:
Adding a test for the config overriding order

weaverryan committed:
Fixing a bug where instanceof was not copied into the new Definition after resolving children

weaverryan committed:
Adding a test for instanceof behavior with parent-child services

This uncovered a bug, where in ResolveDefinitionTemplatePass, we need to be more careful
when setting attributes so that we don't trigger change tracking when there really were
not any changes.

weaverryan committed:
Adding more tests

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/ResolveDefinitionInheritancePass.php:
```diff
@@ -39,57 +39,50 @@ protected function processValue($value, $isRoot = false)
+        $properties = $def->getProperties();
+        foreach ($instanceofDefinition->getProperties() as $k => $v) {
+            // don't override properties set explicitly on the service
+            if (!isset($properties[$k])) {
```
should use array_key_exists, or the `+` operator (would replace the foreach)

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/ResolveDefinitionInheritancePass.php:
```diff
@@ -39,57 +39,50 @@ protected function processValue($value, $isRoot = false)
+        }
+        // append method calls
+        if ($calls = $instanceofDefinition->getMethodCalls()) {
+            $def->setMethodCalls(array_merge($def->getMethodCalls(), $calls));
```
same-name existing calls should not be added IMHO

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Definition.php:
```diff
@@ -803,4 +821,35 @@ public function hasAutowiringType($type)
+        return $this;
+    }
+
+    private function trackChange($type)
```
could be called "touch" (first read did tell me what it did - ie could have been the setter also, with boolean arg)

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php:
```diff
@@ -494,7 +502,13 @@ private function parseDefinition($id, $service, $file, array $defaults)
+                $definition->setTrackChanges(false);
+            }
             $definition->setAutowired($autowire);
+            if ($autowireDefaultsUsed) {
```
"if" not needed

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php:
```diff
@@ -494,7 +502,13 @@ private function parseDefinition($id, $service, $file, array $defaults)
                 throw new InvalidArgumentException(sprintf('Invalid autowire attribute: "by_type", "by_id", true or false expected, "%s" given in "%s".', is_string($autowire) ? $autowire : gettype($autowire), $file));
             }
 
+            if ($autowireDefaultsUsed) {
```
"if" not needed, call `setTrackChanges(!$autowireDefaultsUsed)`

weaverryan committed:
Updating calls logic to merge/override with instanceof

nicolas-grekas reviewed (commented):
This PR implements another merging logic than the current one. If we agree that this is the most intuitive behavior, then I'm fine with it :)

For the reader, the current behavior (pre this PR) is defaults -> service -> instanceof,
my reasoning doing so was from less **explicit** to most explicit:
1. less explicit: I don't know what I'm wiring => defaults
2. a bit more explicit: I know which class I'm wiring => service
3. most explicit: I know the type hierarchy of my class => instanceof

This matches the ordering of the configuration loading (since the code discovers things in this order also - this PR can't change that).

The logic in this PR is from less **specific** to most specific:
1. less specific: all services in that file => defaults
2. a bit more specific: some services in that file => instanceof
3. most specific: *this* service in that file => service

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/ResolveDefinitionInheritancePass.php:
```diff
@@ -39,57 +39,59 @@ protected function processValue($value, $isRoot = false)
+        }
+        // merge method calls
+        if ($instanceofCalls = $instanceofDefinition->getMethodCalls()) {
+            $currentCallMethods = array_map(function($call) {
```
array_column to the rescue

weaverryan replied:
Ah, but now I also need to `strtolower` for the case insensitive matching later :)

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/ResolveDefinitionInheritancePass.php:
```diff
@@ -39,57 +39,59 @@ protected function processValue($value, $isRoot = false)
+
+            $uniqueInstanceofCalls = array_filter($instanceofCalls, function($instanceofCall) use ($currentCallMethods) {
+                // don't add an instanceof call if it was overridden on the service
+                return !in_array($instanceofCall[0], $currentCallMethods);
```
should be case insensitive

weaverryan replied:
+1! Got it + test update

weaverryan committed:
Using Definition instead of ChildDefinition for instanceofConditionals

weaverryan committed:
Fixing outdated type-hint

weaverryan committed:
Fixing method name after rebase

weaverryan committed:
Making instanceof method call matching case insensitive

weaverryan committed:
Fixing an edge case with instanceof property overriding when value is null

weaverryan committed:
A few tweaks thanks to nicolas

weaverryan committed:
Fixing missing use

weaverryan committed:
Thanks fabbot!

weaverryan committed:
Updating more code for ChildDefinition -> Definition change

weaverryan commented:
This is ready to go once again. I updated the description quite a bit to be as clear as possible.

Other than general code review, the question simply comes down to this: how do we want the override logic to work with `_defaults`, `_instanceof` and the service def? @nicolas-grekas explains the 2 here: https://github.com/symfony/symfony/pull/22294#pullrequestreview-31299589

Cheers!

weaverryan committed:
Fixing failing tests: some of the compiler passes caused false "change" positives to the Definition object

weaverryan commented on src/Symfony/Component/DependencyInjection/Compiler/ResolveParameterPlaceHoldersPass.php:
```diff
@@ -58,10 +58,13 @@ protected function processValue($value, $isRoot = false)
             $value->setFile($this->bag->resolveValue($value->getFile()));
             $value->setProperties($this->bag->resolveValue($value->getProperties()));
             $value->setMethodCalls($this->bag->resolveValue($value->getMethodCalls()));
+            $value->setTrackChanges(true);
```
This is a side-effect of the change tracking... changes. This wasn't actually needed to get the new behavior working correctly, it was just needed to get some tests to pass (i.e. the tests were compiling the container and then comparing Definition objects together... one of the Definition objects now had a few extra "changes")

weaverryan commented:
One more update:
1) @nicolas-grekas and I agree that we shouldn't merge/override tags like this, but should update the compiler passes to intelligently handle multiple tags (because sometimes 1 tag should replace another, but other times 2 tags should really stay as 2 tags - i.e. `controller.service_argument`). I will make this change.

2) @nicolas-grekas discussed with me that the new "changes" tracking may be difficult to maintain. Basically, it's not possible for the `service definition config` to override `_instanceof` without this. If we want the behavior described in this PR, it's necessary. If we hate the `changes` too much, then we need to stay with the old logic.

weaverryan committed:
Updating DI tags logic to merge, but not replace tags

The logic is that compiler passes are the only ones that truly have enough information
to know if one tag should replace another, or if both are needed.

weaverryan committed:
thanks fabbot!

GuilhemN commented on src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php:
```diff
@@ -484,9 +484,22 @@ private function parseDefinition($id, $service, $file, array $defaults)
+            }
+            $definition->setAutowired($autowire)
+                ->setTrackChanges(true)
+            ;
```
What about updating this to:
```php
        if (isset($service['autowire'])) {
            $definition->setAutowired($service['autowire']);
        } elseif (isset($defaults['autowire'])) {
            $definition
                ->setTrackChanges(false)
                ->setAutowired($defaults['autowire'])
                ->setTrackChanges(true);
        }

GuilhemN commented on src/Symfony/Component/DependencyInjection/Definition.php:
```diff
@@ -781,4 +798,35 @@ public function hasAutowiringType($type)
+     *
+     * @return $this
+     */
+    public function setTrackChanges($trackChanges)
```
What about marking this as `@internal` ? I don't think we should bother about untracked changes due to a choice of the user.
Maybe we could even use a bound callable where necessary.

GuilhemN commented on src/Symfony/Component/DependencyInjection/Compiler/ResolveDefinitionTemplatesPass.php:
```diff
@@ -95,33 +95,27 @@ private function doResolveDefinition(ChildDefinition $definition)
-        foreach ($definition->getAutowiringTypes(false) as $autowiringType) {
-            $def->addAutowiringType($autowiringType);
+        $parentChanges = $parentDef->getChanges();
+        if (isset($parentChanges['factory'])) {
```
I see one fragility here. 
What if the parent uses defaults?
```yml
services:
    _defaults:
        autowire: true

    parent: ~
```
```yml
services:
    Foo\Bar:
        parent: parent
```
Then `Foo\Bar` won't be autowired, right?

I think the defaults should always be tracked but we should allow to differentiate changes from defaults (maybe just internally, I see no need to do this in userland):
```php
if (isset($defaults['autowire'])) {
    $definition->setTrackChanges('defaults')
        ->setAutowired($defaults['autowire'])
        ->setTrackChanges(true);
}
```
and then use something like the following in the passes needing it:
```php
$definition->getChanges($withoutDefaults = true);
```

GuilhemN replied:
We could also fix https://github.com/symfony/symfony/issues/22345 using this.

weaverryan commented:
Closing in favor of #22356
